Add more error handling and recovery to preload()

tndhagedorn opened the pull request:
An instance of a type stored in Vessel may be un-serializable when
the definition of that type changes.

This can occur when a type's definition is changed but an instance
of that type stored in Vessel is not removed or migrated.

The `preload()` and `preloadBlocking()` APIs can now recover from this state
and continue to preload any remaining valid data.

`preload()`/`prealoadBlocking()` now return the result of the preload operation -
which errors ocurred (if any), and if the preload timeout was hit.

Since this is an API change, this is also a breaking change.

For completeness this same reporting has been added to the profiling data.

Also fixing issue #36

tndhagedorn committed:
Add more error handling and recovery to preload()

An instance of a type stored in Vessel may be un-serializable when
the definition of that type changes.

This can occur when a type's definition is changed but an instance
of that type stored in Vessel is not removed or migrated.

The `preload()` and `preloadBlocking` APIs can now recover from this state
and continue to preload any remaining valid data.

`preload()`/`prealoadBlocking()` now return the result of the preload operation -
which errors ocurred (if any), and if the preload timeout was hit.

Since this is an API change, this is also a breaking change.

For completeness this same reporting has been added to the profiling data.

tndhagedorn committed:
Cleanup

Adding license headers, ensuring these are at top of file

Clean up some comments and spacing

tndhagedorn committed:
Remove some unnecessary temporaries

tndhagedorn committed:
Fix issue #36

If profiling is disabled, profileData should return null

malachid commented on vessel-runtime/src/main/java/com/textnow/android/vessel/VesselImpl.kt:
```diff
@@ -133,46 +131,60 @@ class VesselImpl(
+                        profiler.countBlocking(Event.PRELOAD_TIMEOUT)
                         cursor.close()
-                        return false
+                        return report
```
if we change this to break instead, I think we can remove the cursor.close since cursor.use should cover that use case.

tndhagedorn replied:
Good point.
Given `.use()` always closes the cursor before returning, I don't think it matters whether this is a return or a break.
I did remove the `.close()` though.

malachid commented on vessel-runtime/src/main/java/com/textnow/android/vessel/VesselImpl.kt:
```diff
@@ -185,32 +197,28 @@ class VesselImpl(
 
     override val profileData
-        get() = profiler.snapshot
+        get() = if (DummyProfiler::class.isInstance(profiler)) {
```
it doesn't seem like VesselImpl.kt should know about the different types of Profiler that can be passed in.  Is there a reason the DummyProfiler can not return null for the snapshot?

tndhagedorn replied:
I'd prefer the public contract of `Profiler` to disallow `null` for `snapshot` - I don't see a normal case where a profiler would return a null snapshot, since that's really it's only job.

I added `DummyProfiler` as a way to avoid branching all the VesselImpl code that uses `time()` or `timeBlocking()` when profiling is disabled.

Since the caller needs a way to know whether the snapshot is valid or not (profiling is disabled or not) this getter needs to differentiate somehow.

The other option would be to just always enable profiling and not make it a config option.  Then this is always valid.  It was made optional initially due to concerns raised about performance, but I think we can just leave it on now.  Profiling is enabled for most users today anyways.

malachid commented on vessel-runtime/src/test/java/com/textnow/android/vessel/VesselWrapper.kt:
```diff
@@ -31,11 +55,11 @@ class VesselWrapper(private val vessel: Vessel, private val async: Boolean = fal
     }
 
-    suspend fun preload(timeoutMS: Int?) {
+    suspend fun preload(timeoutMS: Int?): PreloadReport {
```
This syntax can be cleaned up. a single `return if` would be better; or ideally `: PreloadReport = if`

malachid reviewed (commented):
Please update the CHANGELOG and README with the changes.

tndhagedorn committed:
Review comments

Updated CHANGELOG and README
Remove an unnecessary cursor close
Some syntax cleanup

tndhagedorn committed:
Set correct copyright date on new files

tndhagedorn committed:
Fixing typo in README.md
